Sources/SwiftSyntax/WeakLookupTable.swift

```diff
@@ -0,0 +1,228 @@
+    // 0x4000_..., the bucket count must be 0x8000_... thus overflows.
+    precondition(minimalBucketCount <= (Int.max >> 1) + 1)
+
+    var bucketCount = current
+    while bucketCount < minimalBucketCount {
+      // '&*=' for performance. Guaranteed by above 'precondition()'.
+      bucketCount &*= 2
+    }
+    return bucketCount
+  }
+
+  private var _bucketMask: Int {
+    @inline(__always) get {
+      // '&-' for performance. We know 'bucketCount >= 2'.
+      return bucketCount &- 1
+    }
+  }
+
+  @inline(__always)
+  private func _idealBucket(for id: Element.Identifier) -> Int {
+    return id.hashValue & _bucketMask
+  }
+
+  /// Finds the bucket where the object with the specified id should be stored
+  /// to.
+  private
+  func _findHole(_ id: Element.Identifier) -> (pos: Int, alreadyExists: Bool) {
+    var bucket = _idealBucket(for: id)
+
+    // Starting from the ideal bucket for the id, search an available bucket,
+    // or the bucket holding the id.
+    while true {
+      guard let obj = buckets[bucket].value else {
+        return (bucket, false)
+      }
+      if obj.id == id {
+        return (bucket, true)
+      }
+      // '&+' for performance. 'bucketCount' is 0x4000_... or below.
+      bucket = (bucket &+ 1) & _bucketMask
+    }
+  }
+
+  /// Reserves enough space to store the specified number of elements. Returns
+  /// true if resizing happened.
+  func reserveCapacity(_ requiredCapacity: Int) -> Bool {
+    let requiredBucketCount = WeakLookupTable<Element>
+      ._bucketCount(for: requiredCapacity, from: bucketCount)
+    if (bucketCount >= requiredBucketCount) {
+      return false
+    }
+
+    // Slow path. Resizing.
+    let oldBuckets = buckets
+    let oldBucketRange = buckets ..< buckets.advanced(by: bucketCount)
+
+    bucketCount = requiredBucketCount
+    buckets = .allocate(capacity: requiredBucketCount)
+    buckets.initialize(repeating: nil, count: requiredBucketCount)
+
+    // Move all nodes from the old buffer.
+    for oldBucket in oldBucketRange {
+      if let id = oldBucket.pointee.value?.id {
+        let newBucket = buckets.advanced(by: _findHole(id).pos)
+        newBucket.moveAssign(from: oldBucket, count: 1)
+      } else {
+        oldBucket.deinitialize(count: 1)
+      }
+    }
+
+    oldBuckets.deallocate()
+
+    return true
+  }
+
+  /// Count the actual number of occupied buckets.
+  @inline(__always)
+  private func _countOccupiedBuckets() -> Int {
+    var count = 0
+    for i in 0 ..< bucketCount where buckets[i].value != nil {
+      // '&+=' for performance. 'bucketCount' is 0x4000_... or below.
+      count &+= 1
+    }
+    return count
+  }
+
+  /// Reserves enough space to store a single new object. Returns true if
+  /// resizing happened.
+  private func _ensurePlusOneCapacity() -> Bool {
+    // '&+' for performance. 'estimatedCount' is always less than 'bucketCount'
+    // which is 0x4000_... or below.
```

Comment:
This is not true. `estimatedCount` is always less than `buckecCount / maxLoadFactor` (it ends up giving the same guarantees for a `maxLoadFactor > 0.5` but still.

Consider the following example:

```
class Foo: Identifiable {
  let id: Int = 1
}

let lookupTable = WeakLookupTable<Foo>()
lookupTable.reserveCapacity(32)

while true {
  let newValue = Foo()
  lookupTable.insert(newValue)
  print(lookupTable.estimatedCount) /* need to make estimatedCount public for this test */
}
```

With each `insert` `estimatedCount` gets increased until `_minimalBucketCount(for: estimatedCount + 1)`, i.e. `(estimatedCount + 1) / maxLoadFactor` is greater than `bucketCount` at which point the slow path is taken and the lookup table realises that all the previous elements have been released and recounts.

This might influence some of the other comments as well. I haven't looked at them in detail.

Reply:
> `estimatedCount` is always less than `buckecCount / maxLoadFactor` (it ends up giving the same guarantees for a `maxLoadFactor > 0.5` but still.

I believe this is "`estimatedCount` is alway less than or equal to `bucketCount * maxLoadFactor`". So, `'estimatedCount' is always less than 'bucketCount'` is true because `maxLoadFactor` is less than 1.

Either way, what we want to justify here is that `estimatedCount + 1` doesn't overflow. i.e. `estimatedCount < Int.max`, right?

Reply:
Ah, yes. I see. Sorry for the confusion. I somehow assumed that `bucketCount` is `32` in the example above, but its `64`. You're right and it definitely doesn't overflow.